text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,676 @@
+```
+
+Furthermore, if `MyType` is used outside a const context, there are no constness requirements on the
+bounds for types substituted for `T`.
+
+## Drop
+
+A notable use case of `impl const` is defining `Drop` impls.
+Since const evaluation has no side effects, there is no simple example that
+showcases `const Drop` in any useful way. Instead we create a `Drop` impl that
+has user visible side effects:
+
+```rust
+let x = Cell::new(42);
+SomeDropType(&x);
+// x is now 41
+
+struct SomeDropType<'a>(&'a Cell<u32>);
+impl const Drop for SomeDropType {
+    fn drop(&mut self) {
+        self.0.set(self.0.get() - 1);
+    }
+}
+```
+
+You are now allowed to actually let a value of `SomeDropType` get dropped within a constant
+evaluation. This means that
+
+```rust
+(SomeDropType(&Cell::new(42)), 42).1
+```
+
+is now allowed, because we can prove
+that everything from the creation of the value to the destruction is const evaluable.
+
+Note that one cannot implement `const Drop` for structs with fields with just a regular `Drop` impl.
+While from a language perspective nothing speaks against that, this would be very surprising for
+users. Additionally it would make `const Drop` pretty useless. This is explained in more detail in
+the following subsection about generic code and `const Drop`.
+
+```rust
+struct Foo;
+impl Drop for Foo { fn drop(&mut self) {} }
+struct Bar(Foo);
+impl const Drop for Bar { fn drop(&mut self) {} } // not ok
+// cannot call with `T == Foo`, because of missing `const Drop` impl
+fn foo<T: Drop>(t: T) {
+    // Let t run out of scope and get dropped.
+    // Would not be ok if `T` is `Bar`,
+    // because the drop glue would drop `Bar`'s `Foo` field after the `Bar::drop` had been called.
+    // This function is therefore not accept by the compiler.
+}
+```
+
+### const Drop in generic code
+
+`Drop` is special in Rust. You don't need to specify `T: Drop`, but `T::drop` will still be called
+if an object of type `T` goes out of scope. This means there's an implicit assumption, that given
+an arbitrary `T`, we might call `T::drop` if `T` has a drop impl. While we can specify
+`T: Drop` to allow calling `T::drop` in a `const fn`, this means we can't pass e.g. `u32` for
+`T`, because `u32` has no `Drop` impl. Even types that definitely need dropping, but have no
+explicit `Drop` impl (like `struct Foo(String);`), cannot be passed if `T` requires a `Drop` bound.
+
+To summarize, there are currently three ways to interact with `Drop`:
+
+* don't mention `Drop` in the parameter bounds (or mention `?Drop`, amounting to the same thing)
+    * can pass any type that fulfills the other bounds, but may never go out of scope
+* mention `?const Drop` in the parameter bounds
+    * can only pass types with explicit (const or not) `Drop` impls,
+      still can't drop any values inside the function
+* mention `Drop` in the parameter bounds
+    * can only pass types with explicit `const Drop` impls (so no `u32`)
```

Comment:
this bullet seems internally inconsistent (potentially due to a typo?): you say "mention `Drop` in the parameter bounds" for the bullet, but then say in the sub-bullet that such a bound will require explicit `const Drop` impls. I would think that you either want the bullet to say "method `ConstDrop` in the parameter bounds", or you want the sub-bullet to say that the `Drop` bound will only allow passing types with explicit *Drop* (not `const Drop`) impls.

Reply:
This is part of the intro as to why we need `ConstDrop`. The `ConstDrop` trait has not been mentioned before here, and I'm introducing it below to offer a fourth way (and actually useful) way to have `const fn` interact with `Drop`

Having a `T: Foo` bound on a `const fn` will allow you to only pass types that implement `const Foo` if you are within a const context. `Drop` isn't special here, so you can only call a `const fn` with a `T: Drop` bound with a type that has a `const Drop` impl.